Step3.1 View와 ViewController 간 역할 분담 개선

seizze opened the pull request:
### 고민한 점

피드백 주신대로 뷰 초기화 시점에 버튼마다 다른 동작을 넣어놓기 위해서 버튼 종류와 음료수 생성 클로저를 매칭하는 딕셔너리를 선언해 놓고 이용해 봤습니다. BalanceButton과 BeverageLabel도 같은 방법을 적용했습니다.

### 궁금한 점

switch 비교 구문을 없애더라도 tag와 음료수 인스턴스 생성 클로저는 매칭을 해야 할 것 같아서 이것을 딕셔너리로 만들었는데, 새로운 음료수 버튼이 추가될 때 이 딕셔너리가 수정되어야 하긴 해서 이렇게 하는 방법이 맞는지는 잘 모르겠습니다.

그리고 새 버튼이 추가될 때 BeverageButton 클래스의 enum에도 case가 추가되어야 하는데, BeverageButton 클래스가 아예 변경되지 않도록 하기 위해 이 enum도 ViewController에서 갖고 있는 게 나을까요?

seizze committed:
refactor: Change switch case to a dictionary

Change each button to have its own action
Setup button actions at view initialization

seizze committed:
refactor: Remove switch case in balance button

seizze committed:
refactor: Remove switch case in beverage label

seizze committed:
docs: Update README.md

godrm commented on VendingMachineApp/VendingMachineApp/Views/BeverageButton.swift:
```diff
@@ -53,8 +42,8 @@ extension BeverageButton {
-        if let beverage = Item(rawValue: tag)?.action() {
-            customAction?(beverage)
+        if let buttonType = Item(rawValue: tag) {
+            action?(buttonType)
```
다 같은 구조가 됐군요. 만들기 번거로운 것 같지만 CokeButton, FantaButton, StrawberryMilkButton 이런 버튼이 있다면 어떨까요?

seizze replied:
아하 네 그렇게 해보겠습니다!

seizze replied:
BeverageButton을 상속한 각각의 버튼 클래스들을 만들고 각 버튼들이 동작을 갖고 있도록 만들었습니다. 각 버튼 클래스들은 `invokeAction` 메서드를 오버라이드하는 방식으로 동작을 정의하였고, ViewController에서는 이전과 같이 상위클래스인 BeverageButton으로 갖고 있도록 하였습니다!

또한 의도를 명확히 하기 위해 하위 클래스들에 final 키워드도 추가하였습니다.

BalanceButton과 BeverageLabel에도 같은 방법대로 적용하였습니다.

OCP 원칙에 대해서 이론적으로만 알고 있었는데 이번에 진짜로 이해하게 된 것 같습니다!🚀

godrm commented:
다 같은 구조라서 하나만 힌트를 구면 개선할 수 있을 것 같아보이네요.

seizze committed:
refactor: Make beverage button subclasses

Make each subclasses has each actions
Remove view tags and an action dictionary

seizze committed:
refactor: Separate each classes to new files

seizze committed:
refactor: Add subclasses of each balance buttons

seizze committed:
refactor: Add subclasses of each beverage labels

seizze committed:
docs: Update README.md

seizze commented:
피드백 반영하였습니다! ㅎㅎ
